### ISSUE TITLE
HTTP2 CVE's

### ISSUE THREAD

**Issue by blatyo** [ninenines/cowboy#1398]
A number of CVE's were published today that affect a lot of HTTP2 implementations. I wanted to flag them here in case cowboy is also vulnerable to any of them.

Relevant vulnerabilities: https://www.kb.cert.org/vuls/id/605641/
Example of some mitigations: https://github.com/kubernetes/ingress-nginx/commit/333d9fd48d69730a03e087fb6e0f0dae5601c04f

**Comment by essen** [comment]
Well let's go over them one by one. In bold the ones that are more important:

* *Data Dribble*: We should probably have an option to not try to send data unless the window is above a certain size (or the window is larger than what's left to send). This is a good idea even just as an optimization.
* Ping Flood: I'm not too worried about that one but I guess a generic flood detection could be good to have at some point.
* Resource Loop: Cowboy has not implemented the priority mechanism, in part because of issues like this. These have been well known from the start.
* Reset Flood: Same as ping, except here I'd be more worried about the CPU lost to decoding headers (if it goes that far).
* Settings Flood: Same as ping.
* *0-Length Headers Leak*: I suppose that's more relevant if you pre-allocate header strings. For Cowboy that's not the case and they would be an empty binary. Still we probably should reject them since they're not valid, if we don't already.
* *Internal Data Buffering*: I think at some point the `Transport:send` will simply not return and timeout. The question then becomes is the connection dropped as a result. A test would be good to know Cowboy's behavior definitely, regardless of this being used as an attack vector.
* Empty Frame Flooding: Same as ping.

Flood I think we should just count the number of frames received and compare it to the number of non-useful frames received (ping, priority, settings, frames resulting in stream errors, empty frames), if the latter is too high then we drop the connection. Maybe have two options: one indicates how many non-useful frames we accept without question (without considering it a flood even if that's 100% of frames received, can happen if there's no ongoing streams for example), and one indicates the ratio we accept when we look for a potential flood.

**Comment by essen** [comment]
I guess the minimum DATA frame size would be a good default for a minimum window size before we start sending data (if we have less to send we of course just wait for the window to get high enough). Then if this limit is configurable users can increase it depending on their application's needs.

**Comment by essen** [comment]
I suppose we could also be flooded with 1-byte DATA frames, not just empty frames. Or large headers sent with many 1-byte CONTINUATION. I guess a configurable limit in the number of frames received for a specific stream would be a better approach in that scenario. So Empty Frame Flooding is not exactly the same as ping.

**Comment by essen** [comment]
Thoughts welcome on how to detect, fix (if necessary) and test the **Internal Data Buffering** one. I currently have no idea what to do about it.

**Comment by essen** [comment]
Re **Internal Data Buffering**. I think the backpressure implemented recently in master is the fix that was necessary. With it it's possible to configure a maximum size of buffered data for the stream. This with socket options and configurable maximum number of streams should allow limiting the total memory a connection can take. But perhaps it could be augmented with a configurable connection-wide buffered data limit.

I don't really know how to write a test for the vulnerability itself though.

**Comment by essen** [comment]
OK going to add the following options:

* `max_connection_buffer_size` the equivalent of `max_stream_size` but applying to the entire connection, for an easier control of the upper memory bound of a connection. Defaulting to 16MB (stream is 8MB).
* `max_frame_rate` given as `{Intensity, Period}` to configure how many frames are allowed per period of time (including `infinity`). Default probably something like `{1000, 10000}`, 1K frames in 10 seconds. 1K data frames at the default size of 16384 is therefore around 1.6MB/s, which is probably too small for some applications but ought to be large enough in most cases. All frames will be counted because that's just simpler. Rather than use a timeout or query the time every frame I'll just add the frames and check how much time has passed when the maximum has been reached.
* `max_reset_stream_rate` is basically the same except we count the number of streams we've had to reset because an error occurred. The default in this case should be something like `{10, 10000}`, so 10 every 10 seconds. If clients are well written they should not get streams reset.
* `stream_data_threshold` is how large the window ought to be before we start sending data, in order to avoid sending too many DATA frames. Default should be the same as the default frame size, `16384`.

All this plus the 0-length header name fix (where Cowboy will reset the stream) should be good enough to cover all the relevant CVEs. Feedback welcome!

**Comment by essen** [comment]
I've started pushing all this but I'll have to finish tomorrow or Thursday.

**Comment by essen** [comment]
Done.

**Comment by tony612** [comment]
@essen Thanks for the fix. I just feel the default value is a little small for HTTP/2 because users may use multiple streams over one single connection in HTTP/2.

**Comment by essen** [comment]
Could you be more specific? Which default value?

**Comment by tony612** [comment]
I mean max_received_frame_rate. I got an error when I tested sending a large body. But it’s indeed hard to find a perfect default value.

**Comment by essen** [comment]
Maybe it should be increased. Question is to what value. Please open a separate ticket.
